### DISCUSSION TITLE
Can KeyCloak be configured to work with Google CloudSpanner?

### DISCUSSION THREAD

**Discussion by ltoshev** [keycloak/keycloak#31339] [Q&A]
### Description

I have been tasked by our company architects to research the possibility to make KeyCloak work with Google Cloud Spanner.

I see it internally uses Liqubase to setup the DB, and we know there is existing integration between liquibase and cloud spanner. Has anyone done this and can you suggest how can I configure the keycloak docker image to tweak it or configure the liquibase to use cloud spanner. We we the PG dialect with PG-Adapter for spanner.

Resources: 

# https://contribute.liquibase.com/extensions-integrations/directory/database-tutorials/cloud-spanner/
# https://www.keycloak.org/server/db#_configuring_a_database

Right now when I run the image I get this error when its pointed to a CloudSpanner instance:

[org.keycloak.connections.jpa.updater.liquibase.lock.CustomLockService] (main) Failed to create lock table. Maybe other transaction created in the meantime. Retrying...: liquibase.exception.DatabaseException: ERROR: DDL statements are only allowed outside explicit transactions. [Failed SQL: (0) CREATE TABLE public.databasechangeloglock (ID INTEGER NOT NULL, LOCKED BOOLEAN NOT NULL, LOCKGRANTED TIMESTAMP WITHOUT TIME ZONE, LOCKEDBY VARCHAR(255), CONSTRAINT databasechangeloglock_pkey PRIMARY KEY (ID))]

### Motivation

We want to run it in a distributed system where we want to run the KeyCloak in GCP in a multi region setup. That is why we want to use Spanner as its multi region DB .

**Comment by chandragraph8** [comment]
@ltoshev Did you find any answer for this or solved it somehow? I am doing the same.

**Comment by codespearhead** [comment]
Here’s the list of supported databases [1][2], meaning these are the databases Keycloak has been tested with.

Other databases might work if you provide the appropriate JDBC driver [3][4], but issues with this type of setup won’t be addressed unless you start a discussion in the "Ideas" category, explain your use case, and it gains enough support to become a supported database.

Alternatively, you can submit PRs to fix any bugs you encounter with unsupported databases, and as long as you provide the fix yourself, it’s very likely to be merged, as shown here [5].

[1] https://github.com/keycloak/keycloak/blob/98a6ddead0d7121470f5d17be635e1f4dbf3058e/docs/guides/server/db.adoc?plain=1#L14
[2] https://www.keycloak.org/server/db#_supported_databases
[3] https://github.com/keycloak/keycloak/blob/98a6ddead0d7121470f5d17be635e1f4dbf3058e/docs/guides/server/db.adoc?plain=1#L52
[4] https://www.keycloak.org/server/db#_installing_the_oracle_database_driver
[5] https://github.com/keycloak/keycloak/pull/30669

**Comment by ccancellieri** [comment]
+1
